cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -3839,6 +3851,115 @@ if(ARROW_ORC)
+    list(APPEND ARROW_BUNDLED_STATIC_LIBS ${OPENTELEMETRY_LIBRARIES})
+  else()
+    set(OPENTELEMETRY_CMAKE_ARGS ${OPENTELEMETRY_CMAKE_ARGS} "-DWITH_API_ONLY=ON")
+    if(WIN32)
+      # WITH_ETW does not respect WITH_API_ONLY
+      set(OPENTELEMETRY_CMAKE_ARGS ${OPENTELEMETRY_CMAKE_ARGS} -DWITH_ETW=OFF)
+    endif()
+  endif()
+  if(CMAKE_SYSTEM_PROCESSOR STREQUAL "s390x")
+    externalproject_add(opentelemetry_ep
+                        ${EP_LOG_OPTIONS}
+                        URL_HASH "SHA256=${ARROW_OPENTELEMETRY_BUILD_SHA256_CHECKSUM}"
+                        # OpenTelemetry tries to determine the processor arch for vcpkg,
+                        # which fails on s390x, even though it doesn't use vcpkg there
+                        CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env ARCH=s390x
+                                          ${CMAKE_COMMAND} -G ${CMAKE_GENERATOR}
+                                          "<SOURCE_DIR><SOURCE_SUBDIR>"
+                                          ${OPENTELEMETRY_CMAKE_ARGS}
+                        BUILD_COMMAND ${CMAKE_COMMAND} --build "<BINARY_DIR>" --target all
+                        INSTALL_COMMAND ${CMAKE_COMMAND} --build "<BINARY_DIR>" --target
+                                        install
+                        # CMAKE_ARGS ${OPENTELEMETRY_CMAKE_ARGS}
+                        URL ${OPENTELEMETRY_SOURCE_URL}
+                        BUILD_BYPRODUCTS ${OPENTELEMETRY_BUILD_BYPRODUCTS}
+                        EXCLUDE_FROM_ALL NOT
+                        ${ARROW_WITH_OPENTELEMETRY})
+  else()
+    externalproject_add(opentelemetry_ep
+                        ${EP_LOG_OPTIONS}
+                        URL_HASH "SHA256=${ARROW_OPENTELEMETRY_BUILD_SHA256_CHECKSUM}"
+                        CMAKE_ARGS ${OPENTELEMETRY_CMAKE_ARGS}
+                        URL ${OPENTELEMETRY_SOURCE_URL}
+                        BUILD_BYPRODUCTS ${OPENTELEMETRY_BUILD_BYPRODUCTS}
+                        EXCLUDE_FROM_ALL NOT
+                        ${ARROW_WITH_OPENTELEMETRY})
+  endif()
+  add_dependencies(toolchain opentelemetry_ep)
+  add_dependencies(toolchain-tests opentelemetry_ep)
+
+  set(OPENTELEMETRY_VENDORED 1)
+
+  set_target_properties(opentelemetry-cpp::common
+                        PROPERTIES INTERFACE_LINK_LIBRARIES
+                                   "opentelemetry-cpp::api;opentelemetry-cpp::sdk;Threads::Threads"
+  )
+  set_target_properties(opentelemetry-cpp::resources
+                        PROPERTIES INTERFACE_LINK_LIBRARIES "opentelemetry-cpp::common")
+  set_target_properties(opentelemetry-cpp::trace
+                        PROPERTIES INTERFACE_LINK_LIBRARIES
+                                   "opentelemetry-cpp::common;opentelemetry-cpp::resources"
+  )
+
+  foreach(_OPENTELEMETRY_LIB ${_OPENTELEMETRY_LIBS})
+    add_dependencies(opentelemetry-cpp::${_OPENTELEMETRY_LIB} opentelemetry_ep)
+  endforeach()
+endmacro()
+
+# For now OpenTelemetry is always bundled from upstream
+if(1)
```

Comment:
Any chance you can do the same thing that's done elsewhere in this file and call `find_package` if `OpenTelemetry_SOURCE == "SYSTEM"`?

Bundling unconditionally like this makes it really difficult to deal OpenTelemetry provided by a package manager.

Reply:
I can, though without packages in Conda etc. it'll be a little more annoying to test things.

Reply:
It's fine without, maybe we can get that packaged up before the next release.

Reply:
I'm not sure why do we need OpenTelemetry even when we don't enable Flight...

Reply:
This is a failure on my part for not communicating well, sorry.

A few of us working on the C++ query engine/datasets project have been using OpenTelemetry more recently to do some profiling and performance analysis, and so I reoriented the PR along those lines, instead of focusing on Flight. We'll try to get a mailing list post up to explain the thinking before pushing on this more.

The other reason is that I found it difficult to #ifdef OpenTelemetry out completely in those use cases, since fully using it requires you to occasionally pass around context objects and so some internal function signatures changed. Hence it was easier to just enable it unconditionally (but in "api-only" mode). However we may be able to use it without going that far, so for this PR I will try to change it to be conditional.

Reply:
Oh, actually, we did talk about it [a little bit](https://lists.apache.org/thread/2wj87fw2vmokx4pmhm1whr6gt79wmqyb), but that was back in June and it was not obvious that the use case was different. However, I'll still poke the ML, and I'll improve things here so we can build without OpenTelemetry entirely (which would sidestep the RPM issues too, though I'd like to figure that out still) since it'll be more helpful for development (for now).

Reply:
I think at some point we may end up with a new "query engine" module which includes things like the substrait consumer and data holder (these components need to reference datasets and parquet and so can't be placed in the compute module).  The "query engine" module would then include OpenTelemetry not just as a development module but in production too.  I see the ability to generate statistics on running queries to be an essential part of a query engine and not just a development profiling feature.

Maybe then we could enable OT only if "flight" or "query engine" were enabled?  Either way, I agree this is a good discussion for the ML.

Reply:
Oh, sorry. I missed the post.

> Maybe then we could enable OT only if "flight" or "query engine" were enabled? Either way, I agree this is a good discussion for the ML.

+1